### PR TITLE
Add a new trillium-acme crate to automatically manage HTTPS certificates

### PR DESCRIPTION
**Pull request by joshtriplett**
trillium-acme helps you serve HTTPS with Trillium using automatic
certificates, via Let’s Encrypt and ACME tls-alpn-01 challenges.

I've tested this both with the staging and production Let's Encrypt
directories.

### CODE REVIEW

File: acme/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        }
+    };
+
+    return (
+        Acceptor {
+            acme_acceptor,
+            server_config,
+        },
+        future,
+    );
```
[the remark this answers is not in the document]
Comment (joshtriplett) [reply acme/src/lib.rs]: Fixed.

**Comment by joshtriplett** [comment]
Test failure looks unrelated (issue with fastrand not returning reproducible data).

**Comment by joshtriplett** [comment]
Anything further needed for this? Does this seem like a reasonable approach?

**Comment by jbr** [comment]
As a high level concern before I review this more closely:

I'd prefer this not to be in the monorepo, as I'm not sure that I'm the right person to commit to maintaining this for posterity, as I have no experience using acme in production. The goal for trillium is to foster an ecosystem of crates that are maintained by a community of users with a range of needs and use cases. The goal is not for every trillium crate to live within this monorepo or to be maintained by me, and I likely should be more clear about this on the [contributing](https://trillium.rs/contributing) page.

I very much want trillium-acme to exist, and would like to ensure we find a home for it. In order of preference:

1. You publish this under your own name — in practice I will PR anything necessary to track breaking changes and will always trillium-acme tests against changes, possibly even in CI.
2. We publish this in a new trillium-rs repository that you have write access to and primary ownership of, with whatever maintenance expectations set distinctly from trillium itself.
3. We add this to the current repo — as mentioned above, I'd prefer to avoid this as I don't feel that I have the context or expertise to maintain this crate, and as a process matter think it's healthiest for trillium crates to be owned and maintained by their authors

File: acme/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    let mut state = config.state();
+    let server_config = Arc::new(
+        ServerConfig::builder()
+            .with_safe_defaults()
+            .with_no_client_auth()
+            .with_cert_resolver(state.resolver()),
+    );
+    let acme_acceptor = state.acceptor();
+
+    let future = async move {
```
Comment (jbr) [inline acme/src/lib.rs:114]: This future is awkward, and possibly an indication that trillium needs a new affordance to spawn tasks from within init. When does this stream receive items? Is it generally sequenced with the acme_acceptor returning None? Is it fully async and unrelated to the request/response cycle?

Comment (joshtriplett) [reply acme/src/lib.rs:114]: It's fully async, and handles things like renewal.

I would be happy to hide this detail, but 1) I would need a runtime-independent way to spawn a task, and 2) I would need access to the stopper, which also needs to be runtime-independent.

**Comment by joshtriplett** [comment]
> I'd prefer this not to be in the monorepo

I didn't realize this. I kinda liked the trillium monorepo approach. But that said, my primary concern was CI, and if you're offering to build-test trillium-acme in CI, that sounds great.

> I very much want trillium-acme to exist, and would like to ensure we find a home for it. In order of preference:
> 
>     1. You publish this under your own name — in practice I will PR anything necessary to track breaking changes and will always trillium-acme tests against changes, possibly even in CI.
> 
>     2. We publish this in a new trillium-rs repository that you have write access to and primary ownership of, with whatever maintenance expectations set distinctly from trillium itself.

I have some preference towards a trillium-rs repository, because I would want this to live in an organization anyway, and because I would be happy for you to have push access, not just PR access. (Though I'm still happy to maintain it, and I'm *thrilled* that you're willing to PR it for changes to trillium.)

In an ideal world, I would love to find a way to *test* it in CI; that's just a substantial challenge that would require either a CI setup with a domain name pointed to it or a local ACME test server. Should be possible to set something like that up using https://github.com/letsencrypt/pebble .

**Comment by joshtriplett** [comment]
@jbr I've now published `trillium-acme` as a standalone crate, with repository at https://github.com/joshtriplett/trillium-acme . I've invited you as a collaborator.

I managed to implement full CI testing for `trillium-acme`, using https://github.com/letsencrypt/pebble as an ACME server.

**Comment by jbr** [comment]
Thank you for this! I wasn't waiting you out, but was hoping to land an ability to spawn tasks from inside handlers before merging this. Sorry I didn't communicate that. We'll get that in for trillium-acme 0.2!

**Comment by joshtriplett** [comment]
@jbr No problem. I look forward to coordinating further on it. I'd be thrilled if it ends up being either in the trillium-rs organization or in this repo, but I also don't mind maintaining it as a standalone crate. And I do appreciate the idea of integrating the spawning of the future, as well, which will make it much easier to use.

**Pull request merged**